sched/adaptdl_sched/policy/pollux.py

```diff
@@ -311,17 +345,28 @@ def _mutation(self, states, **kwargs):
         prob = 1.0 / np.where(states > 0, num_nonzero, num_zero)
         prob = prob.reshape(states.shape)
         m = np.random.random(states.shape) < prob
-        r = np.random.randint(np.iinfo(np.int16).max, size=states.shape)
+        r = np.random.randint(self._min_replicas, self._max_replicas + 1,
+                              size=states.shape)
         states = (states + m * r) % (self._max_replicas + 1)
+        # Assign minimum allowable replicas for jobs
+        min_replicas_idx = [i for i, j in enumerate(self._jobs)
+                            if j.min_replicas > 0]
+        states[:, min_replicas_idx] = self._min_replicas[min_replicas_idx]
         return states.reshape(states.shape[0], -1)
 
     def _repair(self, pop, **kwargs):
         states = pop.get("X")
         states = states.reshape(states.shape[0], *self._base_state.shape)
-        # Enforce at most one distributed job per node.
+        # Copy previous allocations for non-preemptible jobs
+        states[:, self._nonpreemptible_indices] = \
+            self._base_state[self._nonpreemptible_indices, :]
```

Comment:
Can we simply make sure the mutation does not touch the already-allocated non-preemptible jobs?

Reply:
That fails my unit test. Looks like in some cases `_mutation` is skipped but `_repair` is guaranteed.